File: go/source/vector.go
package source

import (
	"math"
	"math/rand"
)

// Vector represents an Nd vector
type Vector struct {
	x, y, z float64
}

// NewVector creates a vector from a given dimension
func NewVector(x, y, z float64) Vector { return Vector{x, y, z} }

// X returns vector.x
func (vec Vector) X() float64 { return vec.x }

// Y returns vector.y
func (vec Vector) Y() float64 { return vec.y }

// Z returns vector.z
func (vec Vector) Z() float64 { return vec.z }

// XTo sets value of x
func (vec *Vector) XTo(x float64) { vec.x = x }

// YTo sets value of y
func (vec *Vector) YTo(y float64) { vec.y = y }

// ZTo sets value of z
func (vec *Vector) ZTo(z float64) { vec.z = z }

// Equals compares if two vectors are equal
func (vec Vector) Equals(other Vector) bool { return vec == other }

// Add adds two vectors
func (vec Vector) Add(other Vector) Vector {
	return Vector{vec.x + other.x, vec.y + other.y, vec.z + other.z}
}

// Sub subtracts one vector from the other
func (vec Vector) Sub(other Vector) Vector {
	return Vector{vec.x - other.x, vec.y - other.y, vec.z - other.z}
}

// Mul multiplies two vectors
func (vec Vector) Mul(other Vector) Vector {
	return Vector{vec.x * other.x, vec.y * other.y, vec.z * other.z}
}

// Div divides one vector with another vector
func (vec Vector) Div(other Vector) Vector {
	return Vector{vec.x / other.x, vec.y / other.y, vec.z / other.z}
}

// AddS adds a vector to a scalar
func (vec Vector) AddS(other float64) Vector {
	return Vector{vec.x + other, vec.y + other, vec.z + other}
}

// SubS subtracts a scalar from a vector
func (vec Vector) SubS(other float64) Vector {
	return Vector{vec.x - other, vec.y - other, vec.z - other}
}

// MulS multiplies a vector to a scalar
func (vec Vector) MulS(other float64) Vector {
	return Vector{vec.x * other, vec.y * other, vec.z * other}
}

// DivS divides one vector with another vector
func (vec Vector) DivS(other float64) Vector {
	return Vector{vec.x / other, vec.y / other, vec.z / other}
}

// IAdd adds two vectors inplace
func (vec *Vector) IAdd(other Vector) { vec.x += other.x; vec.y += other.y; vec.z += other.z }

// ISub subtracts one vector from the other inplace
func (vec *Vector) ISub(other Vector) { vec.x -= other.x; vec.y -= other.y; vec.z -= other.z }

// IMul multiplies two vectors inplace
func (vec *Vector) IMul(other Vector) { vec.x *= other.x; vec.y *= other.y; vec.z *= other.z }

// IDiv divides one vector with another vector inplace
func (vec *Vector) IDiv(other Vector) { vec.x /= other.x; vec.y /= other.y; vec.z /= other.z }

// IAddS adds a vector to a scalar inplace
func (vec *Vector) IAddS(other float64) { vec.x += other; vec.y += other; vec.z += other }

// ISubS subtracts a scalar from a vector inplace
func (vec *Vector) ISubS(other float64) { vec.x -= other; vec.y -= other; vec.z -= other }

// IMulS multiplies a vector to a scalar inplace
func (vec *Vector) IMulS(other float64) { vec.x *= other; vec.y *= other; vec.z *= other }

// IDivS divides one vector with another vector inplace
func (vec *Vector) IDivS(other float64) { vec.x /= other; vec.y /= other; vec.z /= other }

// Dot is an inner product
func (vec Vector) Dot(other Vector) float64 { return vec.x*other.x + vec.y*other.y + vec.z*other.z }

// L2 means l2 norm
func (vec Vector) L2() float64 { return vec.Dot(vec) }

// Length is the length of a vector
func (vec Vector) Length() float64 { return math.Sqrt(vec.L2()) }

// Unit returns the unit direction of a vector
func (vec Vector) Unit() Vector { return vec.DivS(vec.Length()) }

// Sqrt computes element-wise Sqrt for vector
func (vec Vector) Sqrt() Vector { return Vector{math.Sqrt(vec.x), math.Sqrt(vec.y), math.Sqrt(vec.z)} }

// IsNaN indicates if a vector has an NaN
func (vec Vector) IsNaN() bool { return math.IsNaN(vec.x) || math.IsNaN(vec.y) || math.IsNaN(vec.z) }

// Cross is a cross product for two 3D vectors
func (vec Vector) Cross(other Vector) Vector {
	return Vector{
		vec.y*other.z - vec.z*other.y,
		vec.z*other.x - vec.x*other.z,
		vec.x*other.y - vec.y*other.x,
	}
}

// VectorRandom creates a random vector inside a sphere that has its radius given
func VectorRandom(radius float64, gen *rand.Rand) Vector {
	for {
		random := Vector{gen.Float64(), gen.Float64(), gen.Float64()}
		if random.L2() <= 1 {
			return random.MulS(radius)
		}
	}
}

File: src/main/java/com/app/source/Vector.java
package com.app.source;

import java.util.concurrent.ThreadLocalRandom;

public final class Vector {
    private double x;
    private double y;
    private double z;

    public Vector() {
    }

    public Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public double x() {
        return this.x;
    }

    public double y() {
        return this.y;
    }

    public double z() {
        return this.z;
    }

    public double get(int i) {
        switch (i) {
            case 0:
                return x();
            case 1:
                return y();
            case 2:
                return z();
            default:
        }
        assert false;
        return 0.;
    }

    public void x(double v) {
        this.x = v;
    }

    public void y(double v) {
        this.y = v;
    }

    public void z(double v) {
        this.z = v;
    }

    @Override
    public boolean equals(Object obj) {
        var other = (Vector) obj;

        return this.x == other.x && this.y == other.y && this.z == other.z;
    }

    public Vector add(Vector other) {
        return new Vector(this.x + other.x, this.y + other.y, this.z + other.z);
    }

    public Vector add(double v) {
        return new Vector(this.x + v, this.y + v, this.z + v);
    }

    public Vector sub(Vector other) {
        return new Vector(this.x - other.x, this.y - other.y, this.z - other.z);
    }

    public Vector sub(double v) {
        return new Vector(this.x - v, this.y - v, this.z - v);
    }

    public Vector mul(Vector other) {
        return new Vector(this.x * other.x, this.y * other.y, this.z * other.z);
    }

    public Vector mul(double v) {
        return new Vector(this.x * v, this.y * v, this.z * v);
    }

    public Vector div(Vector other) {
        return new Vector(this.x / other.x, this.y / other.y, this.z / other.z);
    }

    public Vector div(double v) {
        return new Vector(this.x / v, this.y / v, this.z / v);
    }

    public void iadd(Vector other) {
        this.x += other.x;
        this.y += other.y;
        this.z += other.z;
    }

    public void iadd(double v) {
        this.x += v;
        this.y += v;
        this.z += v;
    }

    public void isub(Vector other) {
        this.x -= other.x;
        this.y -= other.y;
        this.z -= other.z;
    }

    public void isub(double v) {
        this.x -= v;
        this.y -= v;
        this.z -= v;
    }

    public void imul(Vector other) {
        this.x *= other.x;
        this.y *= other.y;
        this.z *= other.z;
    }

    public void imul(double v) {
        this.x *= v;
        this.y *= v;
        this.z *= v;
    }

    public void idiv(Vector other) {
        this.x /= other.x;
        this.y /= other.y;
        this.z /= other.z;
    }

    public void idiv(double v) {
        this.x /= v;
        this.y /= v;
        this.z /= v;
    }

    public double dot(Vector other) {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }

    public double l2() {
        return dot(this);
    }

    public double length() {
        return Math.sqrt(l2());
    }

    public Vector unit() {
        return div(length());
    }

    public Vector Sqrt() {
        return new Vector(Math.sqrt(x), Math.sqrt(y), Math.sqrt(z));
    }

    public boolean isNaN() {
        return Double.isNaN(x) || Double.isNaN(y) || Double.isNaN(z);
    }

    public Vector abs() {
        return new Vector(Math.abs(x), Math.abs(y), Math.abs(z));
    }

    public Vector cross(Vector other) {
        return new Vector(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z,
                this.x * other.y - this.y * other.x);
    }

    @Override
    public String toString() {
        return String.format("Vector{%f, %f, %f}", x, y, z);
    }

    public static Vector uniform(double n) {
        return new Vector(n, n, n);
    }

    public static Vector o() {
        return Vector.uniform(0.);
    }

    public static Vector i() {
        return new Vector(1., 0., 0.);
    }

    public static Vector j() {
        return new Vector(0., 1., 0.);
    }

    public static Vector k() {
        return new Vector(0., 0., 1.);
    }

    public static Vector random() {
        var current = ThreadLocalRandom.current();
        return new Vector(current.nextDouble(), current.nextDouble(), current.nextDouble());
    }

    public static Vector randomBall(double radius) {
        for (;;) {
            var random = Vector.random();
            assert 0. <= random.x() && random.x() < 1;
            assert 0. <= random.y() && random.y() < 1;
            assert 0. <= random.z() && random.z() < 1;
            if (random.l2() <= 1.) {
                return random.mul(radius);
            }
        }
    }
}
